### PR TITLE
Initial version of apidoc tool

### PR DESCRIPTION
**Pull request by kalafut**
### Overview 
The goal of this tool is to leverage the OpenAPI specifications that Vault can generate dynamically to improve the [Vault API documentation](https://www.vaultproject.io/api/index.html) . The approach taken is intentionally very selective. One can choose to auto generate docs as many or few API operations as desired, and the result is an in-place update to existing Markdown documentation files. The source of documentation is ultimately Go code, and the state of that code for documentation varies widely. Furthermore, some part of docs may not fit for the prevailing style/template, so it was important to be able to include and exclude on a per operation basis. And because the output is simply an edit to existing Markdown, none of the downstream rendering or published steps change at all.

The generation pipeline is:
1. Write Go code with documentation in Path operations structures 
2. Build Vault
3. Generate a single OpenAPI document  for all backends: `scripts/gen_openapi.sh`
4. Update documentation from OpenAPI specs: `go run scripts/apidoc.go`

A Makefile target called “apidoc” has been added to execute these steps.

### Code Changes
Generating parameter descriptions doesn’t require any code changes. If an operation description block (i.e. the intro text before the table) or payload/request examples are desired, an upgrade to framework.OperationHandler is needed. An example of this is included in the PR. 

### Markdown Changes
To add auto generated docs to an existing Markdown file you add an “apidoc” HTML comment block:

```
<!--apidoc_start path:/pki/config/urls method:post -->
<!--apidoc_end -->
```

This will trigger generation of available docs for the given path and method. Elements will include, if defined: description, table, parameters, payload, request (in that order). One or more of these elements maybe be suppressed using the “disable” tag:

```
<!--apidoc_start path:/pki/config/urls method:post disable:description,payload -->
<!--apidoc_end -->
```

The parameter list order will be in a random initially, as there is no definition of ordering in the code. But edits to the order will be preserved in subsequent generations. Therefore, it makes sense to wrap an existing block of API documentation with the apidoc tags to retain the currently published order.

The `scripts/apidoc.go` command has a couple of options, none of which are required:
* `-f` specifies the source OpenAPI json file (default: openapi.json)
* `-v` verbose mode. Shows all files processed (i.e. that have apidoc tags). Without this only changed files are shown.
* `-r` root for the search for “.md” files to processes (default: website/source/api)
* `-a` analyze. Don’t update any files, but show what percent of any given file is autogenerated (ostensibly to help guide the migration to autogeneration)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -174,6 +174,11 @@ spellcheck:
 	@echo "==> Spell checking website..."
 	@misspell -error -source=text website/source
 
+apidoc: dev
```
Comment (tyrannosaurus-becks) [inline Makefile]: On my computer, I can't get this to work unless I add `export VAULT_ADDR=http://localhost:8200` in `gen_openapi.sh` before mounting all the backends. Might need to unset it at the end of the script, too.

File: website/source/api/secret/pki/index.html.md [hunk 2]
```diff
@@ -318,47 +318,47 @@ $ curl \
...
 
-| Method   | Path                         |
-| :--------------------------- | :--------------------- |
-| `POST`   | `/pki/config/urls`           |
-
-### Parameters
+This endpoint allows setting the issuing certificate endpoints, CRL distribution 				points, and
+OCSP server endpoints that will be encoded into issued certificates. 				You can update any of the
+values at any time without affecting the other 				existing values. To remove the values, simply use
+a blank string as the 				parameter.
```
Comment (tyrannosaurus-becks) [inline website/source/api/secret/pki/index.html.md]: After I ran `make apidoc` on my Ubuntu machine, it edited this file to look like this. Is that expected for the tool? Will it still render just fine or did I do something wrong by re-running it on something that had already been formatted?

File: scripts/apidoc.go [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+	"object":  `{}`,
+}
+
+func main() {
+	flags.Parse(os.Args[1:])
+
+	oas, err := loadOpenAPI(*sourceFile)
+	if err != nil {
+		fmt.Println(err)
+		return
```
Comment (tyrannosaurus-becks) [inline scripts/apidoc.go]: Alternatively, might consider `os.Exit(1)`. Not sure if it really matters.

**Review by tyrannosaurus-becks — COMMENTED** [review]
This looks really good. The only consequential question is the one about the output from giving the tool a spin.

File: website/source/api/secret/pki/index.html.md (hunk 2 above)
Comment (kalafut) [reply website/source/api/secret/pki/index.html.md]: Hmmm, I don't get that. When I ran it just now it undid that change. Can you try again (I just pushed the correctly formatted version)?

File: scripts/apidoc.go (hunk 3 above)
Comment (jescalan) [reply scripts/apidoc.go]: Good call, feels like it does matter to me - getting a proper exit code matters for the netlify build.

**Review by jescalan — COMMENTED** [review]
This is looking really good, super excited to make this happen. I'm sure it will be a great benefit for everyone working on the docs to be able to keep the logic grouped together. Just want to talk about two things we're working on at the web platform level that would interact with this, and throw them out there for consideration and discussion.

1. We are planning to convert markdown files to [MDX](https://mdxjs.com/), which gives us the opportunity to conform to standard markdown syntax, while allowing us to drop in components to introduce unique functionality (for example, our alerts that we use a custom markdown syntax for) that can even be interactive. There may be an opportunity to integrate these two technologies here somehow to make the developer experience smoother.

2. We do feel as if ultimately it would be a benefit to move the code for the website out of this repo, but keep all the documentation here, just to separate concerns a bit more as we grow and add features to the docs sites. Ultimately, vault itself and the website that displays its docs are two separate projects, and it's much easier to handle PRs, team access, and even security concerns if they are treated as such. This is however a feature that increases the direct coupling between the source code and the docs, so we'd have to think about the best way to handle this. We're still at the early stages thinking about what the best way might be to make the website/source code split happen architecturally, but it's good for my team to see this work at least so we can take it into account. And we certainly will keep in sync on how we're thinking about it over time.

File: scripts/apidoc.go (hunk 3 above)
Comment (kalafut) [reply scripts/apidoc.go]: Updated.

**Review by tyrannosaurus-becks — APPROVED** [review]
Awesome! Thanks for putting this together!

**Comment by tyrannosaurus-becks** [comment]
Hi @kalafut ! Should we go ahead and close this issue for now, or merge it? Just trying to clean up PRs that haven't moved in a while. Thanks!

**Comment by kalafut** [comment]
Will push a new PR once updated with all of the docs structure changes.

**Pull request closed without merging**